Zac/add json web token

zbarnard1 opened the pull request:
Adds support for using JWTs in the form of a Bearer token header for authentication.

zbarnard1 committed:
Add Authentication middleware specs

Trying to abide by the Boy Scout Rule to leave things better than I
found them.

zbarnard1 committed:
Add jwt gem

zbarnard1 committed:
Add support for JWTs

We'll allow access to the API using a Bearer token, which will originate
from Amazon Cognito. This will accompany requests as a header,
effectively replacing the P4L-EMAIL header in the authentication
middleware. Verification of the JWTs is handled by the JsonWebToken
class, which relies on the JWKClient to fetch and cache the JWKs for
decoding an incoming JWT.

To simulate the token we'll get from Cognito, the
AuthenticationSpecHelpers now generate a JWT from an RSA key and export
the JWK to a MockJWKClient.

In tests, we always want to be using the MockJWKClient rather than the
real one, so by using `included` in the AuthenticationSpecHelpers we can
automatically stub this out for all tests that include it.

zbarnard1 committed:
Add sub to User

The `sub` is the primary key for the User in the Identity Provider
(IDP), which right now is only Amazon Cognito. This will uniquely
identify every User and we'll use this to allow access to that User's
data in our API. This effectively replaces the User email which we
temporarily used to look up a User.

ittullos commented on spec/tasks/seed_users_spec.rb:
```diff
@@ -1,20 +1,12 @@
-
   it 'creates the users' do
+    expect { SeedUsers.run }.to output(/Inserted 1 users/).to_stdout
+
```
This makes alot more sense than running the rake task in a before block. That code seemed like alot just to run the task.

zbarnard1 replied:
It is, and the line `Rake.application.load_rakefile` was running the task, too!
This is why we like abstractions around certain things. By wrapping the logic of the task in a Ruby class, you can now easily write tests for something that used to be tough to work with, and you decouple your logic from the thing that actually runs it.
You can do it, and there are some great blogs about this. Here's one I really like: https://thoughtbot.com/blog/test-rake-tasks-like-a-boss

Something like this might be good if we wanted to test a task that takes its input from a file on the command line, but for these (and this project) it's definitely overkill.

zbarnard1 replied:
I think this one was the base for a thing I built at Trineo, I used it a ton over the next 3 years so it was worth the effort.
https://medium.com/@jelaniwoods/rspec-tests-for-rake-tasks-da7985896014

zbarnard1 commented:
@ittullos This is ready to go. You can review this commit-by-commit, but the final commit contains all the meat.
I am happy to split it up a little bit if you'd like, but since we've talked through it already it was much easier to structure it in this way.
I'll be the first to point out that there isn't great testing coverage for our JsonWebToken class, but since it mostly delegates to the `jwt` gem or our JWKClient (which has good coverage) I don't think I'm going to lose any sleep over it. We can definitely pair on writing some specs for it to get some practice with testing using stubs.

Let me know what questions you have!

zbarnard1 commented on Gemfile.lock:
```diff
@@ -36,12 +42,22 @@ GEM
     factory_bot (6.4.2)
       activesupport (>= 5.0.0)
+    hashdiff (1.1.1)
+    httparty (0.22.0)
```
OOPS! This never gets used.

ittullos commented:
Everything here looks great dude!! Love how the names of the method calls explain what's going on so well! Made it super easy to follow. I left some comments and questions, hopefully they aren't too nit picky. Just tried to ask a question any time I was curious about something. Thanks again for everything brotha!!

ittullos reviewed (commented):
Looks good! Left a few comments!

zbarnard1 commented on app/lib/json_web_token.rb:
```diff
@@ -0,0 +1,36 @@
+        jwks: jwk_loader
+      }
+    )
+    @payload
```
Great question, no we don't need to handle the error here, it'll bubble up the call stack.

zbarnard1 commented on app/lib/json_web_token.rb:
```diff
@@ -0,0 +1,36 @@
+    @token = token
+  end
+
+  def verify!
```
I'll add some tests for this. The `#verify!` method will raise an error (which comes from `JWT.decode`), which will bubble up the call stack to our authentication middleware where it's rescued.

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -7,15 +7,28 @@ def initialize(app)
+    rescue KeyError, JWT::DecodeError
+      unauthorized_response
+    end
+
```
We _may_ want some error logging, but I don't know just yet. I'll work on deploying to Heroku next, and we'll know a lot more about what additional logging we need at that point.

zbarnard1 commented on app/middleware/authentication.rb:
```diff
@@ -7,15 +7,28 @@ def initialize(app)
+    end
+
+    def extract_token(env)
+      Rack::Request.new(env).fetch_header('HTTP_AUTHORIZATION').split(' ').second
```
Yes, great catch.

zbarnard1 commented on spec/lib/json_web_token_spec.rb:
```diff
@@ -0,0 +1,44 @@
+  subject { described_class.new(token) }
+
+  describe '#verify!' do
+    context 'when the token is valid' do
```
Absolutely, I'll add some.

zbarnard1 replied:
I've added a test that will raise the error we're looking for. I couldn't get it to raise any other kind of error, so I didn't add a bunch of redundant tests but I'm happy to if you'd like more coverage.

ittullos replied:
Nah this is perfect! Thanks dude!

zbarnard1 committed:
fixup! Add support for JWTs

- Adds a test for an token signed by a different RSA key to test error
  raising behavior
- Extends the authentication middleware to check the authorization
  header scheme

ittullos reviewed (approved):
Looks great!!
